service/service.go

```diff
@@ -71,6 +71,19 @@ func newService(set *settings) (*service, error) {
 	if err = srv.telemetryInitializer.init(set.BuildInfo, srv.telemetrySettings.Logger, set.Config.Service.Telemetry, set.AsyncErrorChannel); err != nil {
 		return nil, fmt.Errorf("failed to initialize telemetry: %w", err)
 	}
+
+	// Used to capture main error and any defer errors
+	var returnErrs error
+
+	// If there is any error after we initialize telemetry we need to ensure we clean that up for any future service
+	defer func() {
+		if returnErrs != nil {
```

Comment:
If there is no error, are we inferring that `telemetryInitializer` was properly shutdown? 

If so, it seems like there might be a way to tighten up the way we manage shutting it down. Maybe there is a way to _only_ call it here?

Otherwise, i think it might be cleaner to split `newService` into two functions, where the inner function handles all the cases that can result in a returnErr. Instead, just return from the inner function and check the error in the outer func. Then we don't need to manage this with a defer.

Reply:
I think I'm miss understanding. The logic here is should be if there is an error it will be stored in `returnErrs`. The defer will execute after `returnErrs` is set but before the `return` so if we return due to an error we clean up and if we cleanly return we do nothing and keep the `telemetryInitalizer` running.

Let me know if you still have concerns as I might be missing what you're intending.

Reply:
Reading back through this a few times I think I understand your point. I think there is a way to restructure this function so it doesn't need a defer.